source/common/common/macros.h

```diff
@@ -2,6 +2,14 @@
 
 namespace Envoy {
 
+/**
+ * Assert memory bounds to avoid copy errors.
+ */
+#define SAFE_MEMCPY(dst, src)                                                                      \
+  do {                                                                                             \
+    memmove(dst, src, std::min(sizeof(*(src)), sizeof(*(dst))));                                   \
```

Comment:
what happens if we static_assert that sizes of src and dst are the same?

Reply:
The thing is that there are certain places were reinterpret_cast is used. So, you cannot use static_assert (please look at http://eel.is/c++draft/expr.const#2.15)

As an example, in source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc (around line 205), using static_assert(sizeof(\*(src)) == sizeof(\*(dst)) won’t compile with the current version of c++ because of the pointer *v6*.     
```
pp_ipv6_addr* v6;        
v6 = reinterpret_cast<pp_ipv6_addr*>(&buf[PROXY_PROTO_V2_HEADER_LEN]);        
sockaddr_in6 ra6, la6;        
memset(&ra6, 0, sizeof(ra6));       
memset(&la6, 0, sizeof(la6));
ra6.sin6_family = AF_INET6;
ra6.sin6_port = v6->src_port;
SAFE_MEMCPY(ra6.sin6_addr.s6_addr, v6->src_addr);
```

Reply:
Are you saying that `v6->src_addr` and `ra6.sin6_addr.s6_addr` are different sizes?

That seems surprising.

Maybe this is simply a case where SAFE_MEMCPY is not appropriate, and it needs its own special scrutiny. I think SAFE_MEMCPY should be used when the sizing obviously matches.

Reply:
Not at all, the actual reason for not applying the static_assert is that the comparisson would not be possible since the reinterpret_cast used for initializing v6 is not constexpr. 

Maybe there is a way to determine if src and dst sizes are the same at compile-time without the static_assert, I have not given to much thought to it yet since first I want to compile these few changes before moving on (because I am quiet new to this toolchains 😅 ). Furthermore, other pieces of code may need to be refactored before working with the macro, hence I cannot tell whether changing all the memcpy is the right course of action.

Reply:
Hmmm... I don't see how the compiler can't know about the sizing at compile-time, but just make it an ASSERT then. The only problem with that is if the code isn't covered in tests. Hopefully it is.

Reply:
Ok, making it a C assert is feasible (though errors will be show at runtime), but still, from a pointer, I am not sure if sizeof can take the exact length of the array. As an example, when using strlen to determine the number of bytes in a const char*, most implementations iterate over it until the end (usually marked by the \0 character) is reached.

Maybe, a good idea would be to use another type which included the size attribute, for example, c++ arrays. However, that would increase the overhead in memory allocation. 

Also, I could have missinterpreted the original scope of SAFE_MEMORY and the idea is to make heavy use of MemBlockBuilder 🤔

Reply:
My suggestion was to use Envoy's ASSERT macro, which disappears for release builds, not the one in <assert.h>.

Also I think the static_assert should work. There's nothing variable-length about that type AFAICT. Maybe try expanding out the code by hand without going through the macro to debug it. Then you can see which side of the expression is not constexpr.

Breaking it down, one side is a type defined just above. The other side is from https://man7.org/linux/man-pages/man7/ipv6.7.html
```
           struct sockaddr_in6 {
               sa_family_t     sin6_family;   /* AF_INET6 */
               in_port_t       sin6_port;     /* port number */
               uint32_t        sin6_flowinfo; /* IPv6 flow information */
               struct in6_addr sin6_addr;     /* IPv6 address */
               uint32_t        sin6_scope_id; /* Scope ID (new in 2.4) */
           };

           struct in6_addr {
               unsigned char   s6_addr[16];   /* IPv6 address */
           };
```
So I hope `sizeof(...s6_addr)` is 16, and so is `sizeof(src_addr)`, all discoverable at compile-time. Nothing null-terminated here AFAICT.